Add documentation on configuring caching. (#2199)
cortexproject/cortex: docs/guides/caching.md

---
title: "Caching in Cortex"
linkTitle: "Caching in Cortex"

> weight: 1

Reviewer: I would argue it's not the first item of the menu we want (ie. look better to keep "Running Cortex in Production" as first item):

```suggestion
weight: 5
```

slug: caching
---

> # Caching in Cortex

Reviewer: Please remove this line, otherwise you end up with double title:
```suggestion
```

Ie.
![Screen Shot 2020-03-03 at 10 32 25](https://user-images.githubusercontent.com/1701904/75762006-4bd5a380-5d3a-11ea-9f28-41e1742338d0.png)

Author: How do you get the docs to build locally?

Reviewer: There's some initial setup to be done (ie. install Hugo). There's a basic documentation in `CONTRIBUTING.md`, but I haven't checked how updated it is.

Then:
```
# Run every time you change the doc
./tools/website/web-pre.sh

# Keep it running in another shell
make web-serve
```

Reviewer: _I'm going to retry the local setup and submit a doc PR. Looks me and Cyril are the only one with building the website locally._

Correctly configured caching is important for a production-ready Cortex cluster.

> Cortex has many opportunities for using caching to accelerate queries and reduce cost.  Cortex can use a cache for:

Reviewer: Please add a new line after this line, otherwise the markdown list is not rendered:

![Screen Shot 2020-03-03 at 10 33 02](https://user-images.githubusercontent.com/1701904/75762064-5ee87380-5d3a-11ea-8f4b-9783188c661c.png)

> * Individual chunks

Reviewer: We should make more clear what applies to the chunks storage only and what to any chunk engine.

Ie.
```
Chunks storage engine:

* Individual chunks
* Index lookups for one label on one day
* Reducing duplication of writes

Any storage engine:

* The results of a whole query
```

I would also improve the doc below (when you talk about the caching opportunities) to re-iterate when they applies to the chunks storage engine only.

* The results of a whole query
* Index lookups for one label on one day
* Reducing duplication of writes.

This doc aims to describe what each cache does, how to configure them and how to tune them.

## Cortex Caching Options

> Cortex can use various different technologies for caching - Memcached, Redis or an in-processd FIFO cache.

Reviewer: ```suggestion
Cortex can use various different technologies for caching - Memcached, Redis or an in-process FIFO cache.
```

The recommended caching technology for production workloads is [Memcached](https://memcached.org/).
Using Memcached in your Cortex install means results from one process can be re-used by another.
In-process caching can cut fetch times slightly and reduce the load on Memcached, but can only be used by a single process.

If multiple caches are enabled for each caching opportunities, they will be tiered – writes will go to all caches, but reads will first go to the in-memory FIFO cache, then memcached, then redis.

### Memcached

For small deployments you can use a single memcached cluster for all the caching opportunities – the keys do not collide.

For large deployments we recommend separate memcached deployments for each of the caching opportunities, as this allows more sophisticated sizing, monitoring and configuration of each cache.
For help provisioning and monitoring memcached clusters using [tanka](https://github.com/grafana/tanka), see the [memcached jsonnet module](https://github.com/grafana/jsonnet-libs/tree/master/memcached) and the [memcached-mixin](https://github.com/grafana/jsonnet-libs/tree/master/memcached-mixin).

> Cortex uses DNS discovery to find the various memcached servers in a cluster.

Reviewer: I would better explain we only support DNS service records.

```
-<prefix>.memcache.write-back-buffer int
    How many chunks to buffer for background write back. (default 10000)
-<prefix>.memcache.write-back-goroutines int
    How many goroutines to use to write back to memcache. (default 10)
-<prefix>.memcached.batchsize int
    How many keys to fetch in each batch.
-<prefix>.memcached.consistent-hash
    Use consistent hashing to distribute to memcache servers.
-<prefix>.memcached.expiration duration
    How long keys stay in the memcache.
-<prefix>.memcached.hostname string
    Hostname for memcached service to use when caching chunks. If empty, no memcached will be used.
-<prefix>.memcached.max-idle-conns int
    Maximum number of idle connections in pool. (default 16)
-<prefix>.memcached.parallelism int
    Maximum active requests to memcache. (default 100)
-<prefix>.memcached.service string
    SRV service used to discover memcache servers. (default "memcached")
-<prefix>.memcached.timeout duration
    Maximum time to wait before giving up on memcached requests. (default 100ms)
-<prefix>.memcached.update-interval duration
    Period with which to poll DNS for memcache servers. (default 1m0s)
```

> See [MemcachedConfig documentation](/docs/configuration/config-file-reference.md#memcached_config) if you use a config file with Cortex.

Reviewer: ```suggestion
See [`memcached_client_config ` documentation](../configuration/config-file-reference.md#memcached-client-config) if you use a config file with Cortex.
```

### FIFO Cache (Experimental)

The FIFO cache is an in-memory, in-process (non-shared) cache that uses a First-In-First-Out (FIFO) eviction strategy.
The FIFO cache is useful for simple scenarios where deploying an additional memcached server is too much work, such as when experimenting with the Query Frontend.
The FIFO cache can also be used in front of Memcached to reduce latency for commonly accessed keys.
The FIFO cache stores a fixed number of entries, and therefore it’s memory usage depends on the caches value’s size.

To enable the FIFO cache, use the following flags:

```
-<prefix>.cache.enable-fifocache
    Enable in-memory cache.
-<prefix>.fifocache.duration duration
    The expiry duration for the cache.
-<prefix>.fifocache.size int
    The number of entries to cache.
```

> 

Reviewer: ```suggestion
See [`fifo_cache_config` documentation](../configuration/config-file-reference.md#fifo-cache-config) if you use a config file with Cortex.

```

### Redis (Experimental)

You can also use [Redis](https://redis.io/) for out-of-process caching; this is a relatively new addition to Cortex and is under active development.

```
-<prefix>.redis.enable-tls
    Enables connecting to redis with TLS.
-<prefix>.redis.endpoint string
    Redis service endpoint to use when caching chunks. If empty, no redis will be used.
-<prefix>.redis.expiration duration
    How long keys stay in the redis.
-<prefix>.redis.max-active-conns int
    Maximum number of active connections in pool.
-<prefix>.redis.max-idle-conns int
    Maximum number of idle connections in pool. (default 80)
-<prefix>.redis.password value
    Password to use when connecting to redis.
-<prefix>.redis.timeout duration
    Maximum time to wait before giving up on redis requests. (default 100ms)
```

> 

Reviewer: ```suggestion
See [`redis_config` documentation](../configuration/config-file-reference.md#redis-config) if you use a config file with Cortex.

```

## Cortex Caching Opportunities

### Chunks Cache

The chunk cache stores immutable compressed chunks.
The cache is used by queries to reduce load on the chunk store.
These are typically a few KB in size, and depend mostly on the duration and encoding of your chunks.
The chunk cache is a write-through cache - chunks are written to the cache as they are flushed to the chunk store.  This ensures the cache always contains the most recent chunks.
Items stay in the cache indefinitely.

The chunk cache should be configured on the **ingester**, **querier** and **ruler** using the flags without a prefix.

It is best practice to ensure the chunk cache is big enough to accommodate at least 24 hours of chunk data.

> You can use the follow query (from the [cortex-mixin]()) to estimate the required number of memcached replicas:

Reviewer 2: Mixin needs to be linked?

> You can use the follow query (from the [cortex-mixin](https://github.com/grafana/cortex-jsonnet)) to estimate the required number of memcached replicas:

Reviewer: `follow` > `following` ?

### Chunks Cache

The chunk cache stores immutable compressed chunks.
The cache is used by queries to reduce load on the chunk store.
These are typically a few KB in size, and depend mostly on the duration and encoding of your chunks.
The chunk cache is a write-through cache - chunks are written to the cache as they are flushed to the chunk store.  This ensures the cache always contains the most recent chunks.
Items stay in the cache indefinitely.

The chunk cache should be configured on the **ingester**, **querier** and **ruler** using the flags without a prefix.

It is best practice to ensure the chunk cache is big enough to accommodate at least 24 hours of chunk data.
You can use the following query (from the [cortex-mixin](https://github.com/grafana/cortex-jsonnet)) to estimate the required number of memcached replicas:

```promql
// 4 x in-memory series size = 24hrs of data.
    (
        4 *
        sum by(cluster, namespace) (
                cortex_ingester_memory_series{job=~".+/ingester"}
            *
                cortex_ingester_chunk_size_bytes_sum{job=~".+/ingester"}
            /
                cortex_ingester_chunk_size_bytes_count{job=~".+/ingester"}

>         )

Reviewer 3: Corrent me if Im wrong.
it shouldn't be it?
```
4 *
sum by(cluster, namespace) (cortex_ingester_memory_series{job=~".+/ingester"})
 *
sum by(cluster, namespace) (rate(cortex_ingester_chunk_size_bytes_sum{job=~".+/ingester"}[5m]))
 /
sum by(cluster, namespace) (rate(cortex_ingester_chunk_size_bytes_count{job=~".+/ingester"}[5m]))
```
